Add breadcrumbs bar to editor widget

corneliusludmann opened the pull request:
#### What it does
This commit adds a breadcrumbs bar to the editor widget. It shows the path to the current file and outline information as breadcrumbs. A click of breadcrumbs allows to switch to siblings.

This fixes #5475.

#### How to test
Open an editor shows the breadcrumbs on top of the editor. Folders with files are clickable. A popup allows to select a file that will be opened immediately. Outline information are added to the breadcrumbs as well.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

AlexTugarev commented:
@corneliusludmann please make sure you've signed an ECA https://accounts.eclipse.org/user/eca

corneliusludmann commented:
> @corneliusludmann please make sure you've signed an ECA https://accounts.eclipse.org/user/eca

Currently working on it … ::-)

vitaliy-guliy commented:
@corneliusludmann could you attach few screenshots?

corneliusludmann commented:
> @corneliusludmann could you attach few screenshots?

Sure. This GIF shows the breadcrumbs in action:

![Peek 2019-10-10 22-12](https://user-images.githubusercontent.com/24960040/66708910-e8ca7c80-ed58-11e9-8b54-b0a2639f15c0.gif)

corneliusludmann commented:
**Attention:** I came across with an issue that occurs when opening the Preferences widget after this commit as described here: #6377. I proposed an solution there: #6378. I suggest fixing that issue first and then rebase this PR.

akosyakov commented on packages/editor/package.json:
```diff
@@ -4,10 +4,14 @@
   "description": "Theia - Editor Extension",
   "dependencies": {
     "@theia/core": "^0.11.0",
+    "@theia/filesystem": "^0.11.0",
```
i don't like it :( There were times when one could use an editor standalone, now we need the whole Theia.

Can we move it to the workspace extension? Why do we need outline extension?

corneliusludmann replied:
Done with commit 03eb644.

akosyakov commented:
How can one disable them? Can we have `breadcrumbs.enabled` as in VS Code?

akosyakov commented:
- It does not work like in VS Code: when one click on the part of a path VS Code allows to siblings files, not to chlid as in this PR.
- for `editor-widget.ts` editor, `editor` segment is not clickable at all

akosyakov commented on packages/editor/src/browser/breadcrumbs/breadcrumbs.css:
```diff
@@ -0,0 +1,84 @@
+.theia-breadcrumbs {
```
please make sure to have copyrights in css files as well

akosyakov commented on packages/editor/src/browser/breadcrumbs/breadcrumbs-utils.ts:
```diff
@@ -0,0 +1,96 @@
+ * Traverse upstream (starting with the HTML element `child`) to find a parent HTML element
+ * that has the CSS class `Breadcrumbs.Styles.BREADCRUMB_ITEM`.
+ */
+export function findParentItemHtmlElement(child: HTMLElement): HTMLElement | undefined {
```
please it a class, see https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#di-function-export

akosyakov commented on packages/monaco/src/browser/monaco-outline-contribution.ts:
```diff
@@ -42,20 +41,17 @@ export class MonacoOutlineContribution implements FrontendApplicationContributio
-        });
+
+        // updateOutline and handleCurrentEditorChanged need to be called even when the outline view widget is closed
+        // in order to udpate breadcrumbs.
```
i don't understand it, if an editor is closed there should not be breadcrumbs.

corneliusludmann replied:
The removed condition does not check if an _editor_ is open or closed it checks if the _outline view_ is open or closed. Previously, the outline changes are propagated when the outline view is open only. Now we need the information even when the outline view is closed because the outline view is not the only one that uses this information anymore. The breadcrumbs need the information as well.

What we could do is to check whether the outline view is open _OR_ the breadcrumbs are visible (by the preferences entry `breadcrumbs.enabled`). I don't know if it is really worth it.

akosyakov commented on packages/editor/src/browser/editor-frontend-module.ts:
```diff
@@ -74,4 +76,13 @@ export default new ContainerModule(bind => {
+    bind(BreadcrumbsRendererFactory).toFactory(ctx =>
+        (editor: TextEditor) => {
+            const childContainer = ctx.container.createChild();
+            childContainer.bind('TextEditor').toConstantValue(editor);
```
please use a symbol, if there are duplicate extensions installed, they can pick a bogus instance

corneliusludmann replied:
Done.

akosyakov commented:
Why don't we reuse file and outline tree widgets to show segments? We have already them, have to create new instances and render on click? I would also prefer if we have something like `BreadcrumbService` and `BreadcrumbContribution` in core which are `URI` and DOM event based, so:
- `editor` extension is using it to `open` a `URI` from breadcrumbs, i.e. `BreadcrumbService.open`
- `filesystem` extension implements `BreadcrumbContribution` to open a new file tree widget for the given 
URI
- `workspace` extension overrides `FileSystemBreadcrumbContribution` in order to customize some bits
- `outline` implement `BreadcrumbContribution` to open a new outline tree widget for the given URI if it has `...` as a fragment

lmcbout commented:
**Note: using Theia in the workspace**
 Close all editors, then open package.json


Found some issues when playing with this PR:
```
1- No folders available from the breadcrumbs pull down
2- First click in the editor does not expand the breadcrumbs menu, need to click on another line to expand it
3-  In VSCode, there is "..." appearing at the end when opening the file when there is no selection
    In VSCode: When selecting at the end of the line for "name" , three dots shows for the selection, 
    In Theia, navigating to the Name  keeps the last structure selection, i.e. engines | devDependencies 
    when selecting after the last char of the line, no refresh.
```

VSCode:
![VSCode_breadcrumbs](https://user-images.githubusercontent.com/419051/66847080-c7c48000-ef40-11e9-8802-d6e5129eecf2.gif)

Theia:
![Theia_breadcrumbs](https://user-images.githubusercontent.com/419051/66847095-cf842480-ef40-11e9-9faf-9b626cc15e49.gif)

corneliusludmann commented:
> Found some issues when playing with this PR:

Thank you for your helpful feedback.

> 1- No folders available from the breadcrumbs pull down

Yes, that is a limitation I forgot to mention in the PR description. I currently working on adding the existing file tree view in the popup. I had it planned as the next evolutionary step. Sorry I didn't mention it.

> 2- First click in the editor does not expand the breadcrumbs menu, need to click on another line to expand it

That seems to be a bug in the existing outline view implementation. When you carefully observe the outline view you will see that on first click the outline is rendered without highlighting the current selection. Only after the second click the selection is highlighted. Thus, fixing this will fix this issue as well. I gonna file an issue for that if not already exist.

> 3-  In VSCode, there is "..." appearing at the end when opening the file when there is no selection

Seems to be another limitation I was not aware of. Consider to implement this as well.

> In VSCode: When selecting at the end of the line for "name" , three dots shows for the selection, 
> In Theia, navigating to the Name  keeps the last structure selection, i.e. engines | devDependencies 
> when selecting after the last char of the line, no refresh.

Oh, yes, you're right. Outline breadcrumbs are not removed when there is no selection. Gonna fix this.

lmcbout commented:
>>2- First click in the editor does not expand the breadcrumbs menu, need to click on another line to expand it

>That seems to be a bug in the existing outline view implementation. When you carefully observe the outline view you will see that on first click the outline is rendered without highlighting the current selection. Only after the second click the selection is highlighted. Thus, fixing this will fix this issue as well. I gonna file an issue for that if not already exist.

There might be a bug in the outline view implementation, but I think it is more than that, If you look at the following video, After opening the file, I selected "@types/chai" On the first click, we can see the blue bar on top of the editor and the data being displayed in the outline view, but keeping to select the same item over and over does not amend the breadcrumbs until I select another item or double click on the current selection.
![Theia_breadcrumbs2](https://user-images.githubusercontent.com/419051/66855686-bafc5800-ef51-11e9-827b-efa375aa91f9.gif)

corneliusludmann commented:
> I would also prefer if we have something like `BreadcrumbService` and `BreadcrumbContribution` in core which are `URI` and DOM event based, so:
> 
> * `editor` extension is using it to `open` a `URI` from breadcrumbs, i.e. `BreadcrumbService.open(uri, {x, y})`
> * `filesystem` extension implements `BreadcrumbContribution` to open a new file tree widget for the given
>   URI
> * `workspace` extension overrides `FileSystemBreadcrumbContribution` in order to customize some bits
> * `outline` implement `BreadcrumbContribution` to open a new outline tree widget for the given URI if it has `...` as a fragment

I refactored the code with commit 03eb644 which introduces `BreadcrumbService` and `BreadcrumbContribution`. Fell free to make your comments.

_I will **squash** the commits once I implemented all suggested changes._

corneliusludmann commented:
> How can one disable them? Can we have `breadcrumbs.enabled` as in VS Code?

Done with commit 3dee5ec.

_I will **squash** the commits once I implemented all suggested changes._

corneliusludmann commented:
> Why don't we reuse file and outline tree widgets to open URIs? We have already them, have to create new instances and render on click?

For the file tree widget, I currently working on it. For the outline tree widget it seems not possible yet because the tree is not easily reusable yet.

corneliusludmann commented:
> > In VSCode: When selecting at the end of the line for "name" , three dots shows for the selection,
> > In Theia, navigating to the Name  keeps the last structure selection, i.e. engines | devDependencies
> > when selecting after the last char of the line, no refresh.
> 
> Oh, yes, you're right. Outline breadcrumbs are not removed when there is no selection. Gonna fix this.

Should be fixed with commit 03eb644.

corneliusludmann commented:
> > > 2- First click in the editor does not expand the breadcrumbs menu, need to click on another line to expand it
> 
> > That seems to be a bug in the existing outline view implementation. When you carefully observe the outline view you will see that on first click the outline is rendered without highlighting the current selection. Only after the second click the selection is highlighted. Thus, fixing this will fix this issue as well. I gonna file an issue for that if not already exist.
> 
> There might be a bug in the outline view implementation, but I think it is more than that, If you look at the following video, After opening the file, I selected "@types/chai" On the first click, we can see the blue bar on top of the editor and the data being displayed in the outline view, but keeping to select the same item over and over does not amend the breadcrumbs until I select another item or double click on the current selection.

Hmm. At first, when I look at your video, the moment when the breadcrumbs appear is in sync with the highlighting of the tree item in the outline view. Therefore, I still think it is an general issue with the outline view. Second, when I try to reproduce your steps I realized that it does not have to do with a single or double click, it is just a delay. When I open an editor, click once, wait a few seconds and do a second click the breadcrumbs appear immediately (same with the highlighting of the outline tree item). Sometimes it needs more than a few seconds. However, in my tests, it is _always_ in sync with the outline view highlighting.

akosyakov commented on packages/core/src/browser/breadcrumbs/breadcrumbs-service.ts:
```diff
@@ -0,0 +1,73 @@
+     * Subscribe to this event emitter to be notifed when the breadcrumbs have changed.
+     * The URI is the URI of the editor the breadcrumbs have changed for.
+     */
+    get onBreadcrumbsChange(): Event<URI> {
```
`onDidChangeBreadcrumbs` or `onBreadcrumbsChanged`

corneliusludmann replied:
Done

akosyakov commented on packages/core/src/browser/breadcrumbs/breadcrumbs-service.ts:
```diff
@@ -0,0 +1,73 @@
+     * Notifies that the breadcrumbs for the given URI have changed and should be re-rendered.
+     * This fires an `onBreadcrumsChange` event.
+     */
+    breadcrumbsChanges(uri: URI): void {
```
firing events should be an implementation detail, i.e. `BreadcrumbsContribution` should have `onDidChange` instead

corneliusludmann replied:
Maybe I don't get it. But it sounds that this would make it more complicated without any substantial benefit since the breadcrumbs renderer has to subscribe to all breadcrumb contributions then.

akosyakov commented on packages/core/src/browser/breadcrumbs/breadcrumbs-contribution.ts:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * The breadcrumb type. Breadcrumbs returned by `#computeBreadcrumbs(uri)` should have this as `Breadcrumb#type`.
+     */
+    type: symbol;
```
`readonly` ? the same about `priority`

corneliusludmann replied:
Done

akosyakov commented on packages/core/src/browser/breadcrumbs/breadcrumbs-service.ts:
```diff
@@ -0,0 +1,73 @@
+    /**
+     * Opens a popup for the given breadcrumb at the given position. `parent` is used as the host element for the newly created popup element.
+     */
+    async openPopup(breadcrumb: Breadcrumb, position: { x: number, y: number }, parent: HTMLElement): Promise<BreadcrumbPopup | undefined> {
```
just `open`?

akosyakov commented on packages/core/src/browser/breadcrumbs/breadcrumb-popup-container-renderer.tsx:
```diff
@@ -0,0 +1,122 @@
+import { BreadcrumbPopup } from './breadcrumb-popup';
+
+export const BreadcrumbPopupContainerRenderer = Symbol('BreadcrumbPopupContainerRenderer');
+export interface BreadcrumbPopupContainerRenderer {
```
Do we need interface, symbol and a class? Should not `BreadcrumbPopupContainerRenderer` be enough? Please see https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#classes-over-interfaces

corneliusludmann commented:
I updated this PR with the file tree for file breadcrumbs and a '…' breadcrumb for outline breadcrumbs. There are still some minor flaws with the outline breadcrumbs that are in sync with flaws in the outline itself. These need to be fixed in the outline implementation. (E.g. selecting a non-collapsed outline element does not collapse the tree in the outline view and does not set the correct breadcrumb.)

![Peek 2019-11-12 12-16](https://user-images.githubusercontent.com/24960040/68667447-47e10400-0546-11ea-9352-5a2ee4333c16.gif)

From my point of view, this PR is ready for another review.

akosyakov commented:
- [ ] We should register and use breadcrumb colors to be aligned with VS Code: https://code.visualstudio.com/api/references/theme-color#breadcrumbs

lmcbout commented:
I was looking to get a second review, But there are merge conflicts :(

akosyakov commented:
- [ ] Breadcrumb should be contributed to the toolbar of a tab bar group and be updated based on the current widget.

paulalexwilson commented:
I was excited by this PR. Do we plan to merge it at some point or is it no-longer being considered?

lmcbout commented:
I am interested by this PR as well, but there was some unresolved and merging issues , they are not resolved yet

corneliusludmann commented:
I could resolve the merge conflicts and rebase the branch. However @akosyakov raised two requests that need to be implemented before merging:
- https://github.com/eclipse-theia/theia/pull/6371#issuecomment-563277489
- https://github.com/eclipse-theia/theia/pull/6371#issuecomment-564104327

Currently I do not have the time to take care of it. If someone else is willing to contribute to this work feel free to push to this branch or to fork this branch. I'm happy to provide assistance when needed.

colin-grant-work commented:
@corneliusludmann, my colleague @kenneth-marut-work and I were thinking of finishing this PR up, unless you were planning on returning to it?

corneliusludmann commented:
@colin-grant-work, unfortunately, I don't. I would be very happy if you would take over the work on this feature.
